drivers/serial/serial.c

```diff
@@ -1370,24 +1326,69 @@ static ssize_t uart_readv(FAR struct file *filep, FAR struct uio *uio)
 #endif
 
   nxmutex_unlock(&dev->recv.lock);
-  if (recvd >= 0)
+  return recvd;
+}
+
+static ssize_t uart_read(FAR struct file *filep,
+                         FAR char *buffer, size_t buflen)
+{
+  return uart_read_internal(filep, buffer, buflen, false);
+}
+
+/****************************************************************************
+ * Name: uart_readv
+ ****************************************************************************/
+
+static ssize_t uart_readv(FAR struct file *filep, FAR struct uio *uio)
+{
+  FAR const struct iovec *iov = uio->uio_iov;
+  int iovcnt = uio->uio_iovcnt;
+  bool nonblock = !!(filep->f_oflags & O_NONBLOCK);
+  ssize_t ntotal;
+  ssize_t nread;
+  int i;
+
+  for (i = 0, ntotal = 0; i < iovcnt; i++)
     {
-      uio_advance(uio, recvd);
+      nread = uart_read_internal(filep, iov[i].iov_base,
+                                 iov[i].iov_len, nonblock);
+      if (nread < 0)
+        {
+          if (ntotal > 0)
+            {
+              break;
+            }
+
+          return nread;
+        }
+
+      ntotal += nread;
+
+      /* Check for a parital success condition, including an end-of-file */
+
+      if (nread < iov[i].iov_len)
+        {
+          break;
+        }
+
+      /* set nonblock flag after first read */
+
+      nonblock = true;
     }
 
-  return recvd;
+  return ntotal;
 }
 
 /****************************************************************************
- * Name: uart_writev
+ * Name: uart_write
  ****************************************************************************/
 
-static ssize_t uart_writev(FAR struct file *filep, FAR struct uio *uio)
+static ssize_t uart_write(FAR struct file *filep, FAR const char *buffer,
```

Comment:
do we apply the similar nonblock change to uart_write

Reply:
I think there is no similar issue with write, only read.